src/serializer/functions.js

```diff
@@ -141,16 +128,38 @@ export class Functions {
         let value = property.descriptor && property.descriptor.value;
         invariant(value !== undefined);
         invariant(value instanceof Value);
-        let entry = this.__optimizedFunctionEntryOfValue(value);
+        let entry = this._optimizedFunctionEntryOfValue(value);
         if (entry) recordedAdditionalFunctions.push(entry);
       }
     }
     return recordedAdditionalFunctions;
   }
 
+  _generateOptimizedFunctionsFromRealm(): Array<AdditionalFunctionEntry> {
+    let realm = this.realm;
+    let recordedAdditionalFunctions = [];
+    for (let [valueToOptimize, argModel] of realm.optimizedFunctions) {
+      let value = valueToOptimize instanceof AbstractValue ? this._unwrapAbstract(valueToOptimize) : valueToOptimize;
+      invariant(value instanceof ECMAScriptSourceFunctionValue);
+      // Check for case where __optimize was called in speculative context where effects were discarded
+      if (!value.isValid()) {
+        let error = new CompilerDiagnostic(
```

Comment:
What still not clear to me: Is this really meant to be a user-facing "error", even if recoverable? The message seems to indicate that it is completely safe to just ignore. At best, this would be "Informational".